src/red.ts

```diff
@@ -129,36 +131,55 @@ export const serializedRedEnvSourceText = (function redEnvFactory(blueEnv: Membr
         if (typeof blue === 'function') {
             return getRedFunction(blue);
         }
-        let isBlueArray = false;
-        try {
-            isBlueArray = isArrayOrNotOrThrowForRevoked(blue);
-        } catch {
-            // blue was revoked - but we call createRedProxy to support distortions
-            return createRedProxy(blue);
-        }
-        if (isBlueArray) {
-            return getRedArray(blue);
-        } else if (typeof blue === 'object') {
+        if (typeof blue === 'object') {
             const red: RedValue | undefined = WeakMapGet(blueMap, blue);
-            if (isUndefined(red)) {
-                return createRedProxy(blue);
+            if (!isUndefined(red)) {
+                return red;
+            }
+            let hasDynamicMark: boolean = false;
+            try {
+                hasDynamicMark = hasOwnPropertyCall(blue, LockerLiveValueMarkerSymbol);
+            } catch {}
+            if (hasDynamicMark) {
+                return createDynamicRedProxy(blue);
+            }
+            let isBlueArray = false;
+            try {
+                isBlueArray = isArrayOrNotOrThrowForRevoked(blue);
+            } catch {
+                // blue was revoked - but we call createRedProxy to support distortions
+                return createStaticRedProxy(blue);
             }
-            return red;
+            if (isBlueArray) {
+                return getStaticRedArray(blue);
+            }
+            return createStaticRedProxy(blue);
         } else {
             return blue as RedValue;
         }
     }
 
-    function getRedArray(blueArray: BlueArray): RedArray {
-        const b: RedValue[] = map(blueArray, (blue: BlueValue) => getRedValue(blue));
-        // identity of the new array correspond to the inner realm
-        return [...b];
+    function getStaticBlueArray(redArray: RedArray): BlueArray {
+        return map(redArray, (red: RedValue) => blueEnv.getBlueValue(red));
+    }
+
+    function getStaticRedArray(blueArray: BlueArray): RedArray {
+        // TODO: static arrays have no identity at this point. If the
+        // same array is sent twice, you get two distinct red arrays. 
+        return map(blueArray, (blue: BlueValue) => getRedValue(blue));
```

Comment:
this is not new, this is what is happening today, just to highlight it so we can address it.